### ISSUE TITLE
How to reduce the seams with current model?

### ISSUE THREAD

**Issue by ps003** [Insta360-Research-Team/DiT360#22]
Is there anything that can be done with the model to reduce the seam?

Also, I didn't realize there is a trigger word. Is it panorama? Asking since I looked at the example prompt and that's the main keyword.

Here's the created image. Notice the seam.

Image
Oh, and I did this on a mac and converted the inference.py to mps for mac testing. :) I'll include all.

[mac_converted_to_mps_v001.py](https://github.com/user-attachments/files/25117404/mac_converted_to_mps_v001.py)

In addition, I'm also testing the https://github.com/cedarconnor/ComfyUI-DiT360 project trying to get to the level of skybox.blockadelabs.com

Fascinating work by your teams!

**Comment by fenghora** [comment]
Thanks for bringing this up, and for the interest in our work!

To reduce the seam, one possible direction is to adjust the training procedure to explicitly supervise the model at the two horizontal boundaries, which we plan to address in our upcoming work.
Concretely, pad the noise right after sampling and include the padded regions in the loss, instead of discarding them during loss computation.
Here is the example code:
```
def training_step(self, batch, batch_idx):

        def latent_circular_pad(x: torch.Tensor, pad_n=2) -> torch.Tensor:
            """
            x: (bsz, c, h, w)
            Return: (bsz, c, h, w+4) by concatenating [x_last2, x, x_first2] along width.
            """
            if x.ndim != 4:
                raise ValueError(f"Expected 4D tensor (bsz, c, h, w). Got shape={tuple(x.shape)}")
            if x.size(-1) < 2:
                raise ValueError(f"Width must be >= 2. Got W={x.size(-1)}")

            x0 = x[..., :pad_n]      # first two columns: (bsz, 16, 64, 2)
            x_1 = x[..., -pad_n:]    # last two columns:  (bsz, 16, 64, 2)

            y = torch.cat([x_1, x, x0], dim=-1)  # (bsz, 16, 64, 132)
            return y

        pixel_latents = encode_images(batch["pixel_values"], self.vae).to(self.dtype)       # (bsz, 16, 64, 128) 
        noise = torch.randn_like(pixel_latents, device=self.device, dtype=self.dtype)
        padding_n = 4
        pixel_latents = latent_circular_pad(pixel_latents, pad_n=padding_n)                 # (bsz, 16, 64, 136) 
        noise = latent_circular_pad(noise, pad_n=padding_n)

        bsz = pixel_latents.shape[0]
        u = compute_density_for_timestep_sampling(
            weighting_scheme=self.weighting_scheme,
            batch_size=bsz,
            logit_mean=self.logit_mean,
            logit_std=self.logit_std,
            mode_scale=self.mode_scale,
        )

        indices = (u * self.noise_scheduler_copy.config.num_train_timesteps).long()
        timesteps = self.noise_scheduler_copy.timesteps[indices].to(device=self.device)

        sigmas = get_sigmas(
            self.noise_scheduler_copy,
            timesteps,
            device=self.device,
            dtype=self.dtype,
            n_dim=pixel_latents.ndim
        )

        noisy_model_input = (1.0 - sigmas) * pixel_latents + sigmas * noise

        packed_noisy_model_input = DiT360Pipeline._pack_latents(
            noisy_model_input,
            batch_size=bsz,
            num_channels_latents=noisy_model_input.shape[1],
            height=noisy_model_input.shape[2],
            width=noisy_model_input.shape[3],
        )   # torch.Size([1, 64 * 136 // 4, 64])

        ##############################################################
        packed_height = noisy_model_input.shape[2] // 2     
        packed_width = noisy_model_input.shape[3] // 2     
        dim = packed_noisy_model_input.shape[-1]          

        latent_image_ids = DiT360Pipeline._prepare_latent_image_ids(
            bsz,
            packed_height,
            packed_width - padding_n,  # remove padding
            self.device,
            self.dtype,
        ) 

        latent_image_ids = latent_image_ids.reshape(packed_height, packed_width - padding_n, 3)
        first_col_image_ids = latent_image_ids[:, :padding_n // 2, :]
        last_col_image_ids = latent_image_ids[:, -padding_n // 2:, :]
        latent_image_ids = torch.cat([last_col_image_ids, latent_image_ids, first_col_image_ids], dim=1)
        latent_image_ids = latent_image_ids.reshape(packed_height * packed_width, 3)
        ##############################################################

        if self.flux_transformer.config.guidance_embeds:
            guidance_vec = torch.full(
                (bsz,),
                self.hparams.args.guidance_scale,
                device=self.device,
                dtype=self.dtype,
            )
        else:
            guidance_vec = None

        prompt_embeds = batch["prompt_embeds"]
        pooled_prompt_embeds = batch["pooled_prompt_embeds"]
        text_ids = batch["text_ids"]

        model_pred = self.flux_transformer(
            hidden_states=packed_noisy_model_input,
            timestep=timesteps / 1000,
            guidance=guidance_vec,
            pooled_projections=pooled_prompt_embeds,
            encoder_hidden_states=prompt_embeds,
            txt_ids=text_ids,
            img_ids=latent_image_ids,
            return_dict=False,
        )[0]


        model_pred = DiT360Pipeline._unpack_latents(
            model_pred,
            height=noisy_model_input.shape[2] * self.vae_scale_factor,
            width=noisy_model_input.shape[3] * self.vae_scale_factor,
            vae_scale_factor=self.vae_scale_factor,
        )

        weighting = compute_loss_weighting_for_sd3(
            weighting_scheme=self.weighting_scheme, sigmas=sigmas)

        target = noise - pixel_latents 

        ############################################################################################

        loss = torch.mean((weighting.float() * (model_pred.float() - target.float()) ** 2).reshape(target.shape[0], -1), 1).mean()

        self.log("train_loss", loss, on_step=True, on_epoch=True,prog_bar=True, logger=True, sync_dist=True)

        return loss
```

Regarding the “trigger word”: you can simply describe it explicitly in the prompt, e.g., “This is a panorama. The image shows …”. 

Happy to discuss further if you have more findings or questions.
